src/Hopper/Internal/LoweredCore/ClosureConversion.hs

```diff
@@ -0,0 +1,281 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Hopper.Internal.LoweredCore.ClosureConversion
+  ( closureConvert
+  ) where
+
+import Hopper.Internal.LoweredCore.ANF
+import Hopper.Internal.LoweredCore.ClosureConvertedANF
+import Hopper.Internal.Type.BinderInfo (BinderInfo(..))
+import Hopper.Internal.Type.Relevance (Relevance(..))
+import Hopper.Utils.LocallyNameless
+
+import Control.Arrow (second)
+import Control.Lens (Lens', Traversal', (^.), (%~), (%=), (?=), _head, at,
+                     firstOf, makeLenses, over, set, use)
+import Control.Monad.State.Class (MonadState, get)
+import Control.Monad.Trans.State.Strict (State, runState)
+import Data.Foldable (toList)
+import Data.Maybe (fromMaybe)
+import Data.Word (Word32)
+
+import qualified Data.DList as DL
+import qualified Data.Map.Strict as Map
+import qualified Data.Vector as V
+
+--------------------------
+-- Data Types
+--------------------------
+
+-- | For tracking in 'EnvState' whether we're building a closure or thunk. We
+-- need to differentiate to decide whether a variable we encounter reaches
+-- outside of the closure or thunk; closures have an extra binding level for
+-- explicit arguments to the function.
+data ClosureType
+  = Thunk
+  | Closure
+  deriving (Eq, Show)
+
```

Comment:
awesome! 
(and amusingly we'll eventually have linear thunks, which *will be used*, its just a question of when they get forced :) )